### PR TITLE
Review docs for 0.2

### LINKED ISSUE #973: Review docs for 0.2
We dedicated a bit of time to check that the docs are still consistent after the many breaking changes.

- [ ] update (untested) platform creation
  - the proper solution is tracked by #984, but it should not be solved necessarily in 0.2 - but the code has to be correct

### PR DESCRIPTION
**Pull request by stavros11**
Close #973.

TODO:
- [x] index
- [x] getting-started
- [x] main-documentation
- [x] tutorials
- [ ] api-reference

In a9f8506 I removed some outdated figures, including the xmind diagram of the `Platform`, because it is no longer true for 0.2. It looked nice though, we could probably add an updated version in some later 0.2.* release.

@alecandido I have been redefining the platform in the getting-started example and having some testing (see #984) would indeed be useful. If you agree and it doesn't take much effort we could even do here.

### CODE REVIEW

**Comment by stavros11** [comment]
@alecandido I went through the main-documentation part of the docs and tried to update to the 0.2 changes. However, I have the impression that this part is redundant and it would be simpler to just expand the docstrings, which are already rendered as part of the API reference. We could potentially move some of text from main-documentation to the docstrings. We could do this directly here or leave as it is for now and update in a later release.

Also, it seems that `doctest` is failing in the CI, while it is passing locally. It may be an issue with py3.12 but I haven’t tried to debug it yet. I’m sure you’ll have different behavior on Mac as well.

I’ll go over the tutorials next, but feel free to have a look in my changes in `index` and `main-documentation`.

**Comment by alecandido** [comment]
> Also, it seems that `doctest` is failing in the CI, while it is passing locally. It may be an issue with py3.12 but I haven’t tried to debug it yet. I’m sure you’ll have different behavior on Mac as well.


```py
Exception raised:
    Traceback (most recent call last):
      File "/opt/hostedtoolcache/Python/3.12.5/x64/lib/python3.12/doctest.py", line 1368, in __run
        exec(compile(example.source, filename, "single",
      File "<doctest python[0]>", line 7, in <module>
        from qibolab.instruments.qm import Octave, QmConfigs, QmController
      File "/home/runner/work/qibolab/qibolab/src/qibolab/instruments/qm/__init__.py", line 2, in <module>
        from .controller import *
      File "/home/runner/work/qibolab/qibolab/src/qibolab/instruments/qm/controller.py", line 11, in <module>
        from qm import QuantumMachinesManager, SimulationConfig, generate_qua_script
    ModuleNotFoundError: No module named 'qm'
```
https://github.com/qiboteam/qibolab/actions/runs/10723526912/job/29741963394?pr=1028#step:10:572

The problem is that `qm-qua` is not yet available in any release for py3.12

**Comment by alecandido** [comment]
> @alecandido I went through the main-documentation part of the docs and tried to update to the 0.2 changes. However, I have the impression that this part is redundant and it would be simpler to just expand the docstrings, which are already rendered as part of the API reference. We could potentially move some of text from main-documentation to the docstrings. We could do this directly here or leave as it is for now and update in a later release.

Yes, I was not aiming for a deep rework of the docs in this release, but we need something.

#984 is part of the game, since we need to better test what we're writing (and separate Python files, to be then embedded in the docs, would be actually easier) and we also need to keep docs minimal, organized, and meaningful.
I would postpone more time-consuming discussions, but just slightly: we could face the issue as soon as we release.

**Comment by alecandido** [comment]
#1021 is also relevant for the comprehensive docs review process.

**Comment by stavros11** [comment]
@alecandido I completed a first pass through the docs, except the API reference, fixing the points I found to be outdated to the 0.1 interface, even though it is very likely I missed something. If you like, you can also have a look.

Regarding the API reference, it is mostly generated automatically from the docstrings, which I would not update as part of this PR. We could go over them separately, however I am not sure if we have time to do that for 0.2.0 (I wouldn't block the release for that), or we should do incrementally later, potentially also moving the parts from main-documentation there, as I proposed above. 

The only issue about API reference that I would consider fixing here is the fact that several objects are repeated, if they are exposed to multiple modules. For example, `Platform` appears at least three times: once in the main `qibolab` package (as `qibolab.Platform`) and twice in the `qibolab.platform` package (as `qibolab.platform.Platform` and `qibolab.platform.platform.Platform`). We should decide which we want to keep, most likely either the uppermost or lowermost, and drop the rest to avoid repetitions. This may also be affected by #1030.

> The problem is that `qm-qua` is not yet available in any release for py3.12

You are right, thanks. I used QM as the getting-started platform example, since this is the only driver known to be working in 0.2. In d12c3811588e36166f326ddc51a4b89366f68794 I disabled doctest for py3.12 as a temporary solution. In >=0.2.1 where we will have more drivers, we could use another instrument in the example, or if we are lucky qm-qua will support py3.12 soon, so we can just lift the restriction from the workflow.

**Comment by alecandido** [comment]
> The only issue about API reference that I would consider fixing here is the fact that several objects are repeated, if they are exposed to multiple modules. For example, `Platform` appears at least three times: once in the main `qibolab` package (as `qibolab.Platform`) and twice in the `qibolab.platform` package (as `qibolab.platform.Platform` and `qibolab.platform.platform.Platform`). We should decide which we want to keep, most likely either the uppermost or lowermost, and drop the rest to avoid repetitions. This may also be affected by #1030.

Yes, I would face this in #1030, since Sphinx should honor the `__all__` attribute (let's see what will happen with the package/subpackage re-exports...).


> @alecandido I completed a first pass through the docs, except the API reference, fixing the points I found to be outdated to the 0.1 interface, even though it is very likely I missed something. If you like, you can also have a look.
> 
> Regarding the API reference, it is mostly generated automatically from the docstrings, which I would not update as part of this PR. We could go over them separately, however I am not sure if we have time to do that for 0.2.0 (I wouldn't block the release for that), or we should do incrementally later, potentially also moving the parts from main-documentation there, as I proposed above.

I will take a look, but the api-reference is fully generated, so I would keep improving it together with the rest of the package, and avoid making a dedicated effort to review it all.
Since it was directly attached to what we developed, I hope it was already the most up-to-date (before this PR), and we'll keep working on it.

> You are right, thanks. I used QM as the getting-started platform example, since this is the only driver known to be working in 0.2. In [d12c381](https://github.com/qiboteam/qibolab/commit/d12c3811588e36166f326ddc51a4b89366f68794) I disabled doctest for py3.12 as a temporary solution. In >=0.2.1 where we will have more drivers, we could use another instrument in the example, or if we are lucky qm-qua will support py3.12 soon, so we can just lift the restriction from the workflow.

I see your point, but I would also consider using dummy. QM is a more realistic example, but it's not much more useful than dummy if you don't own a QM cluster (and you're going to use just that).
In any case, your solution is perfectly acceptable by me. It is a temporary one, but fully satisfactory. So, we can postpone any improvement :)

File: src/qibolab/platform/platform.py
Comment (alecandido) [inline src/qibolab/platform/platform.py:229]: Great, this is making the call truly minimal!
```py
platform.execute([sequence, ...])
```

**Review by alecandido — APPROVED** [review]
I didn't really check all the changes line by line, especially those in `lab.rst`. I really believe we should find a better way to write these docs and examples, but it's not the goal of this PR, we have #984 for that.

We could even do it here, as you proposed, but I'd rather move forward towards the 0.2 release, and postpone #984 at least to the next one.

File: doc/source/tutorials/calibration.rst
Comment (stavros11) [reply doc/source/tutorials/calibration.rst]: Indeed, these weren't available when I wrote this and they would cause `doctest` to fail. Now they are, so I used them in some examples in 75f12e8416c579396fd677bc2f3a453c85996c35.

File: doc/source/tutorials/compiler.rst
Comment (stavros11) [reply doc/source/tutorials/compiler.rst:44]: I was also not sure whether I should update or drop this. I updated since it didn't require much effort and in the end kept it, since, even though we agreed to not make this part of the "public API" I believe that what is written here actually works. So unless we are expecting to break it in the future (in some 0.2.*), it wouldn't hurt and shouldn't require maintenance either. I am also okay with dropping though.

File: doc/source/tutorials/instrument.rst
Comment (stavros11) [reply doc/source/tutorials/instrument.rst:27]: If I remember correctly, while working on this, I opened another branch on top to remove `Instrument.setup()`. However, I eventually gave up mainly due to it being used in https://github.com/qiboteam/qibolab/blob/ff05ddbd8a0a30ccdc2c2e36b05b73489156d2dc/src/qibolab/platform/platform.py#L279-L281

which is usually needed for setting up external LO parameters (eg. for TWPA pumps) and in other similar cases. 

At this stage, I believe `setup()` is only needed for "auxilliary" instruments, not `Controller`s, but I'd be more confident saying it after seeing Qblox in the new structure. Because I believe, `setup` was never actually used by any instrument other than Qblox and LO-like ones. There may even be a way to get rid of it completely, but for the interest of time and other open issues I didn't put much effort.

File: doc/source/index.rst
Comment (stavros11) [reply doc/source/index.rst:17]: Indeed, but this looked more like a summary of what qibolab provides in general, rather than how it is supposed to be used, so I left it as it is, just moving the "pulse API" higher, since this and the platform are the major points. If you see the next bullets it is even talking about the compiler (not even part of the public API) and circuit execution, which are not the main points of qibolab itself, but are still there in terms of code.

Feel free to change something if you like though.

File: src/qibolab/platform/platform.py
Comment (alecandido) [reply src/qibolab/platform/platform.py:228]: Another option is to accept `ExecutionParameters` attributes, as `.execute()` arguments, either exclusively
```py
def execute(self, sequences, ..., **kwargs):
    options = ExecutionParameters(**kwargs)
```
or as a further option
```py
def execute(self, sequences, options = None, ..., **kwargs):
    if options is None:
        options = ExecutionParameters(**kwargs)
```

In case, I'd prefer the first alternative, since "there is only one way", and then it's less confusing (maybe it's not that intuitive to imagine that, in the second case, if you specify explicitly `options`, the other parameters will be ignored).

However, I'm not 100% sure it's just better than the current alternative. Though it would be often less redundant.

Comment (stavros11) [reply src/qibolab/platform/platform.py:228]: Over the two, I would also prefer the first and I believe it also works in the case where no `kwargs` are given. 

From a usability point of view it is probably simpler, since we don't even need to expose `ExecutionParameters` itself in the public API (one less thing for the user to search for/remember).

File: doc/source/tutorials/compiler.rst
Comment (alecandido) [reply doc/source/tutorials/compiler.rst:44]: ~~If it was not for the update, I would have a clear reply: drop it, and then we will resurrect as soon as the `Compiler` will be exposed.~~

At a second thought: we could always search for it with Git, so I'd say drop it, and whenever we will re-release `Compiler` as part of the public API we will search for the updated file (I have done it seldom, but I should be able to find it within Git's history - i.e. without scanning manually commit by commit).

File: doc/source/tutorials/instrument.rst
Comment (alecandido) [reply doc/source/tutorials/instrument.rst:27]: Ok, I see the point: while with controller we have a communication through `play()`, the other instruments would be otherwise configured only upon creation, and this would be a bit unpractical (e.g. to change the frequency of an LO, after platform instantiation).

In this sense, it could even be useful in the long term, but I'd suggest not making it abstract: let's keep the default implementation empty, such that it will be only implemented by those instruments for which it has a non-trivial action.

**Comment by alecandido** [comment]
I would decouple the README rewriting from this, since they have two rather different goals: I'd not aim to actually refactor the docs, we should have just made sure that it is not completely outdated.

Instead, the README is much shorter, and updating it is not truly a major effort, and extremely relevant to present the current state of Qibolab. So, I would take a moment at the very end to decide about a final version (thinking on it in the meanwhile). That's why I opened #1043 to track it

File: doc/source/tutorials/compiler.rst
Comment (stavros11) [reply doc/source/tutorials/compiler.rst]: File removed in 8e23cb994ebb11a4cf0d60002548393cc7e4ea7a. I did the removal alone in a commit, so I believe, if needed, we should be able to recover the file just by reverting this commit.

File: doc/source/tutorials/instrument.rst
Comment (stavros11) [reply doc/source/tutorials/instrument.rst]: > Ok, I see the point: while with controller we have a communication through `play()`, the other instruments would be otherwise configured only upon creation, and this would be a bit unpractical (e.g. to change the frequency of an LO, after platform instantiation).

Indeed, the main reason we cannot do `setup` during creation is that it requires communication to the physical instrument in order to "upload" the settings, however during creation we are not yet connected. I believe we decided to do it this way to postpone connection as much as possible so that we are able to cover more code in the CI.

> In this sense, it could even be useful in the long term, but I'd suggest not making it abstract: let's keep the default implementation empty, such that it will be only implemented by those instruments for which it has a non-trivial action.

`setup` was converted from abstract to an empty method in 7209fef4548eab1fdcbea6a85738a860765ae465. Currently it is only implemented by the `LocalOscillator` instrument.

**Comment by stavros11** [comment]
@alecandido thanks for the comments. I believe most discussions have now been resolved. There are only two left. I don't think any particular action is needed for the first. About the second (`ExecutionParameters`), you could probably implement your solution in #1030, since it is also affects the public API (we could make `ExecutionParameters` fully internal), or otherwise in another small PR. Or we could even do nothing and leave it as it is.

> So, I would take a moment at the very end to decide about a final version (thinking on it in the meanwhile). That's why I opened #1043 to track it

Sure, it should be fast and I would even do after #1030 in order to use the proper imports.

**Comment by alecandido** [comment]
I agree with your summary. Please, proceed to merge whenever you wish

**Pull request merged**
